src/exampleWorker.js

```diff
@@ -0,0 +1,229 @@
+  postMessage('WORKER STARTED');
+  var grammar;
+
+  var examplePieces = {};
+  var generator;
+
+  var overrides = {
+    ident: function(g, ex, syn, args) {
+      return {needHelp: true};
+    },
+    ctor: function(g, ex, syn, args) {
+      return {needHelp: true};
+    },
+    letter: function(g, ex, syn, args) {
+      if (Math.floor(Math.random() * 2) === 0) {
+        return {example: String.fromCharCode(Math.floor(26 * Math.random()) + 'a'.charCodeAt(0))};
+      } else {
+        return {example: String.fromCharCode(Math.floor(26 * Math.random()) + 'A'.charCodeAt(0))};
+      }
+    }
+  };
+
+  self.addEventListener('message', function(e) {
+    postMessage(e.data.name);
+    switch (e.data.name) {
+      case 'initialize':
+        /* eslint-disable no-eval */
+        grammar = ohm.makeRecipe(eval(e.data.recipe));
+        /* eslint-enable no-eval */
+
+        start();
+        break;
+      case 'request:examples':
+        var ruleName = e.data.args[0];
+
+        var examplesForRule = examplePieces[ruleName] || null;
+        self.postMessage({name: 'received:examples',
+                          args: [ruleName, examplesForRule]});
+        break;
+      case 'update:neededExamples':
+        var neededExamples = utils.difference(
+          Object.keys(grammar.rules),
+          Object.keys(examplePieces)
+        );
+
+        self.postMessage({name: 'received:neededExamples',
+                          args: [neededExamples]});
+        break;
+      case 'add:userExample':
+        ruleName = e.data.args[0];
+        var example = e.data.args[1];
+        generator.processExampleFromUser(example, ruleName);
+        break;
+    }
+  }, false);
+
+  function addPiecesToDict(trace, examples) {
+    if (trace.expr.constructor.name === 'Terminal') {
+      return;
+    } else {
+      if (trace.expr.constructor.name === 'Apply') {
+        var ruleName = trace.expr.toString();
+        if (!examples.hasOwnProperty(ruleName)) {
+          examples[ruleName] = [];
+        }
+
+        if (!examples[ruleName].includes(trace.source.contents)) {
+          examples[ruleName].push(trace.source.contents);
+        }
+      }
+      trace.children
+           .filter(function(child) { return child.succeeded; })
+           .forEach(function(child) { return addPiecesToDict(child, examples); });
+    }
+  }
+
+  function ExampleGenerator(examplePieces) {
+    this.examplePieces = examplePieces;
+    this.rules = initialRules(grammar);
+    this.examplesNeeded = utils.difference(
+      Object.keys(grammar.rules),
+      Object.keys(examplePieces)
+    );
+    this.currentRuleIndex = 0;
+  }
+
+  ExampleGenerator.prototype.processExampleFromUser = function(example, optRuleName) {
+    if (optRuleName) {
+      var trace = grammar.trace(example, optRuleName);
+      if (trace.succeeded) {
+        addPiecesToDict(trace, examplePieces);
+      }
+    }
+
+    utils.objectForEach(grammar.rules, function(ruleName) {
```

Comment:
I don't think you need a helper here...wouldn't `Object.keys(grammar.rules).forEach` do the trick?

I find it's a fine balance as to whether these kinds of things make the code more or less readable. For common stuff like this, my opinion is that it's better to be explicit...and in this case, it doesn't even save any characters. When I see something like this, it makes me think there's a reason that you're not using `Object.keys().forEach` -- so I have to go look at another file to confirm it's doing what I think it's doing. If you just use the obvious thing, then I don't have to think about it at all.